src/backend/compatIO.js

```diff
@@ -0,0 +1,60 @@
+// @flow
+
+import fs from "fs-extra";
+import stringify from "json-stable-stringify";
+import {type Compatible} from "../util/compat";
+
+/**
+ * Utility functions, which allow us to read/write a Compatible of any
+ * type to a JSON file. They should automatically detect the Flow types
+ * and produce errors when misused.
+ *
+ * Note, the Compatible<T> type used cannot be opaque.
+ */
+
+/**
+ * Creates a Compatible JSON file writer.
+ * The writer's Flow type for `data` will automatically match the provided
+ * `toJSON`'s argument type.
+ * Optionally takes a `typeName` to improve error messages.
+ */
+export function compatWriter<T, J>(
+  toJSON: (T) => Compatible<J>,
+  typeName?: string
```

Comment:
Rather than take the `typeName` as a string, how about taking the corresponding `COMPAT_INFO`?

Reply:
Well, I haven't found a good way to get Flow types as strings at runtime. So we could get the `COMPAT_INFO` by any-casting the opaque type _after_ `toJSON` succeeded.

Reply:
Oh, by taking you meant manually as an argument? That could work, though I think it's nice to add your own names in free form. Since it's usage is to enhance error messages. It should be something clear to the user.